### PR TITLE
Feature/rate limiting inbound traffic [pr]

### PR DESCRIPTION
**Pull request by rapala61**
#### What's this PR do?
It adds a rate limiter that will protect Gambit member route from 🤖abuse.

#### How should this be reviewed?
- 👀 
- All tests should pass.

#### Relevant tickets
Fixes Pivotal [Card#](https://www.pivotaltracker.com/story/show/160629007)

#### Checklist
- [ ] Create monitoring alert for `member route Rate Limiter (Limit reached)`.
- [ ] Documentation added for new features/changed endpoints.
- [x] Tests added/updated for new features/bug fixes.
- [ ] ENV variables added/updated for new features/bug fixes.
- [ ] Tested on staging.
- [ ] Pinged a PM if this is a larger PR that would benefit from some additional testing love.

### CODE REVIEW

File: lib/cluster.js [hunk 1]
```diff
@@ -47,10 +52,21 @@ function initWorkerProcess() {
...
 }
 
+/**
+ * initMasterRateLimiter
+ * @see https://github.com/animir/node-rate-limiter-flexible/wiki/Cluster#ratelimitercluster
+ */
+function initMasterRateLimiter() {
+  // Doesn't require any options, it is only storage and messages handler
+  // It's a singleton
+  return new RateLimiterClusterMaster();
```
[the remark this answers is not in the document]
Comment (rapala61) [reply lib/cluster.js]: Thanks! It was also setup this way so it was more testable so I added a test for `initMasterRateLimiter`.

File: lib/helpers.js [hunk 2]
```diff
@@ -34,6 +35,19 @@ module.exports.sendErrorResponseWithNoRetry = function (res, err) {
...
 
+module.exports.sendRateLimitedResponse = function (res, rateLimiterRes = {}) {
+  const error = new RateLimitedError();
+  // sets rate limiting headers
+  res.set({
+    'Retry-After': rateLimiterRes.msBeforeNext / 1000,
+    // TODO: 'X-RateLimit-Limit': points,
+    // points are theoretically unique per rate-limiter so this should be injected
+    'X-RateLimit-Remaining': rateLimiterRes.remainingPoints,
+    'X-RateLimit-Reset': new Date(Date.now() + rateLimiterRes.msBeforeNext),
```
Comment (DFurnes) [inline lib/helpers.js]: Are these used by Blink (or Twilio)? Or output into logs somewhere for debugging?

File: test/integration/app/v2-routes/messages/member.test.js [hunk 3]
```diff
@@ -168,3 +183,55 @@ test('POST /api/v2/messages?origin=twilio outbound message should match sendInfo
...
+
+  // mock user fetch twice
+  integrationHelper.routes.northstar
+    .intercept.fetchUserByMobile(inboundRequestPayload.From, member, rateLimiterUpperBound + 1);
+
+  // mock user update twice
+  integrationHelper.routes.northstar
+    .intercept.updateUserById(member.data.id, member, rateLimiterUpperBound + 1);
+
+  [...Array(rateLimiterUpperBound).keys()].forEach(async () => {
```
Comment (DFurnes) [inline test/integration/app/v2-routes/messages/member.test.js]: Is there an advantage to spreading the array returned by `.keys()` into a new array?

File: lib/helpers.js [hunk 4]
```diff
@@ -34,6 +35,19 @@ module.exports.sendErrorResponseWithNoRetry = function (res, err) {
...
+  });
+  return exports.sendErrorResponseWithNoRetry(res, error);
```
[the remark this answers is not in the document]
Comment (rapala61) [reply lib/helpers.js]: Yes, it's applied globally to the Twilio inbound route so any inbound texts from the same user are counted towards the rate limit. **The user doesn't receive any message back to their mobile if they are being rate limited**. Twilio inbound texts are sent to Blink. Blink is the one that will receive the `429` error and message. Blink will ignore this request. However, we [expose logging when the user is rate limited](https://github.com/DoSomething/gambit/blob/feature/rate-limiting-inbound-traffic/lib/middleware/messages/member/rate-limiter.js#L29). This way we can easily detect instances where this is occurring.

I'll increase the upper bound to `25` requests per hour to be safe. Does that sound good? @DFurnes

[the remark this answers is not in the document]
Comment (DFurnes) [reply lib/helpers.js]: Ok, yeah that sounds good to me. Will be good to keep an eye on!

File: lib/helpers.js (hunk 2 above)
Comment (rapala61) [reply lib/helpers.js]: I'll expose them in newRelic as custom attributes of the transaction when logging any errors.

File: test/integration/app/v2-routes/messages/member.test.js (hunk 3 above)
Comment (rapala61) [reply test/integration/app/v2-routes/messages/member.test.js]: In this case we would get an iterator from the `.keys()` method so I wouldn't be able to use the good old `.forEach` unless I spread it inside a new `Array`.

Comment (DFurnes) [reply test/integration/app/v2-routes/messages/member.test.js]: A ha! 💡

**Pull request merged**